### PR TITLE
Add new background illumination methods

### PR DESCRIPTION
**Pull request by starcalibre**
This PR adds fast illumination method used by Broad [[1](http://onlinelibrary.wiley.com/doi/10.1111/jmi.12178/full)] for their images, and better handles images with non-standard bitdepths in 16-bit containers. The new background illumination method find the median filtered version of the mean image, rather than finding the mean of a the median filtered images. This speeds things up significantly.
- `find_background_illumination` now finds the illumination correction field by finding the mean of the related images, then smoothing that out with a rank filter. Default parameter for the filter is p0=0.5 (i.e. a median filter).
- `find_background_illumination` now takes an input bit-depth parameter. This should be used when bit-depths of 12 or 14 bits were captuted by the camera, but a 16-bit TIF is used as a container. This ensures the original bit-depth is preserved in any subsequent steps of the illumination correction. 
- `find_background_illumination` has a default radius of one quarter the width or height the first image, whichever is smaller. After lots of experimentation with BBBC021, BBBC022 and the Cellomics data, I found this rule of thumb gave the best smooth illumination correction field across all channels. It's probably something that needs to be tweaked for every data-set but I'm happy with this default.
- `find_background_illumination` scales images from floats to `uint16` before being passed to the percentile filter. In the case where the mean image had a very limited dynamic range, the resultant smoothed, floating poiint image had quite a large granularity between values. This resulted in artefacts being added to the image during illumination correction. This makes rank filtering slower, but it doesn't matter because this PR reduces the number of rank filterings being done by several orders of magnitude. :)
- `correct_multiimage_illumination` no longer scales all images down to 8-bit, the dtype is now taken from the first image passed to the function.

### CODE REVIEW

File: microscopium/preprocess.py [hunk 1]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
-    # This function follows the "PyToolz" streaming data model to
-    # obtain the illumination estimate. First, define each processing
-    # step:
-    read = io.imread
+    read = tlz.partial(io.imread)
+
+    if input_bitdepth is None:
+        in_range = "image"
+    else:
+        in_range = (0, np.power(2, input_bitdepth) - 1)
```
Comment (jni) [inline microscopium/preprocess.py]: `2**input_bitdepth - 1`

File: microscopium/preprocess.py [hunk 2]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
     --------
     `correct_image_illumination`, `correct_multiimage_illumination`.
     """
-    # This function follows the "PyToolz" streaming data model to
-    # obtain the illumination estimate. First, define each processing
-    # step:
-    read = io.imread
+    read = tlz.partial(io.imread)
```
Comment (jni) [inline microscopium/preprocess.py]: I have actually since seen that the toolz community uses tz, not tlz. I don't want you to clutter this PR with a huge rename/refactor, but maybe add an extra import, "import cytoolz as tz", and use that here?

Comment (jni) [inline microscopium/preprocess.py]: Also, why are you using partial here?

File: microscopium/preprocess.py [hunk 3]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
+        in_range = "image"
+    else:
+        in_range = (0, np.power(2, input_bitdepth) - 1)
+
+    if radius is None:
+        # get default radius from input image
+        im0 = mio.imread(fns[0])
+        radius = np.round(np.min(im0.shape) / 4).astype(np.uint16)
+        # ensure radius is odd
+        radius = radius - np.mod(radius, 2) + 1
```
Comment (jni) [inline microscopium/preprocess.py]: `radius += 1 - radius%2`

Comment (jni) [inline microscopium/preprocess.py]: Keep related variables together. Radius is not needed until *after* the illumination field calculation, so create it there, so that `in_range` is where it is needed — next to the definition of rescale.

File: microscopium/preprocess.py [hunk 4]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
+
+    rescale = tlz.partial(exposure.rescale_intensity, in_range=in_range)
```
Comment (jni) [inline microscopium/preprocess.py]: What is the default out_range?

Comment (jni) [inline microscopium/preprocess.py]: I would prefer if we stated it explicitly.

File: microscopium/preprocess.py [hunk 5]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
     normalize = (tlz.partial(stretchlim, bottom=stretch_quantile)
                  if stretch_quantile > 0
                  else skimage.img_as_float)
```
Comment (jni) [inline microscopium/preprocess.py]: Are you still doing the stretchlim? Should we simply remove this step?

File: microscopium/preprocess.py [hunk 6]
```diff
@@ -549,39 +557,58 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
                          'not recognised.' % method)
 
-    return illum
+    # apply median filter to find ICF
+    pad = tlz.partial(util.pad, pad_width=radius, mode='reflect')
+    rank_filter = fun.partial(rank.percentile, selem=skmorph.disk(radius),
+                              p0=quantile)
+    _unpad = tlz.partial(unpad, pad_width=radius)
+
+    # ignore the loss of precision warning
```
Comment (jni) [inline microscopium/preprocess.py]: good stuff

File: microscopium/preprocess.py [hunk 7]
```diff
@@ -631,8 +663,8 @@ def correct_multiimage_illumination(im_fns, illum, stretch_quantile=0,
     for im in ims_pass2:
         corrected = im / illum
         rescaled = exposure.rescale_intensity(corrected, in_range=corr_range,
-                                              out_range=np.uint8)
-        out = np.round(rescaled).astype(np.uint8)
+                                              out_range=im0.dtype.type)
+        out = np.round(rescaled).astype(im0.dtype)
```
Comment (jni) [inline microscopium/preprocess.py]: This is a very nice improvement. However, I'm concerned about memory blowout — do we really need anything more than 8 bits after the illumination correction?

File: tests/test_pre.py [hunk 8]
```diff
@@ -12,21 +12,21 @@
...
+    j = np.array([[1, 10, 0, 9, 0],
+                  [3, 10, 4, 1, 1],
+                  [4, 10, 0, 7, 4],
+                  [9, 3, 2, 0, 7],
+                  [1, 3, 3, 9, 3]], np.uint8)
+    k = np.array([[9, 1, 7, 7, 3],
+                  [9, 1, 6, 2, 2],
+                  [2, 8, 2, 0, 3],
+                  [4, 3, 8, 9, 10],
+                  [6, 0, 2, 3, 10]], np.uint8)
```
Comment (jni) [inline tests/test_pre.py]: ech! revert these changes please.

File: tests/test_pre.py [hunk 9]
```diff
@@ -59,11 +59,11 @@ def test_illumination_median(image_files):
...
-    illum_true = np.array([[ 4.,  5.,  4.,  1.,  1.],
-                           [ 4.,  6.,  2.,  4.,  2.],
-                           [ 8.,  3.,  4.,  2.,  7.],
-                           [ 3.,  3.,  3.,  7.,  6.],
-                           [ 3.,  3.,  3.,  7.,  7.]]) / 10
+    illum_true = np.array([[4., 5., 6., 2., 2.],
+                           [5., 5., 2., 6., 2.],
+                           [4., 3., 6., 4., 7.],
+                           [3., 3., 3., 7., 7.],
+                           [4., 3., 2., 6., 6.]]) / 10
```
Comment (jni) [inline tests/test_pre.py]: Have you manually verified these results?

**Comment by jni** [comment]
A few comments to address. Also, it looks like you need to cover a few more of the code paths!

https://coveralls.io/builds/6008472/source?filename=microscopium%2Fpreprocess.py#L560

File: microscopium/preprocess.py (hunk 2 above)
Comment (starcalibre) [inline microscopium/preprocess.py]: LOL I have no idea. It's a bit useless isn't it.. :P

File: microscopium/preprocess.py (hunk 4 above)
Comment (starcalibre) [inline microscopium/preprocess.py]: The [default](http://scikit-image.org/docs/dev/api/skimage.exposure.html#skimage.exposure.rescale_intensity) it based on the images' dtype, but sure I'll put that in explicitly. It'll make the code a bit easier to follow.

File: microscopium/preprocess.py (hunk 5 above)
Comment (starcalibre) [inline microscopium/preprocess.py]: Only if a ``stretch_quantile`` is passed to the function? Now that everything is based on the image's bit-depth I can't see any reason to do contrast stretching at this this stage. I'm happy to remove it.

File: microscopium/preprocess.py (hunk 7 above)
Comment (starcalibre) [inline microscopium/preprocess.py]: That's an interesting question ;) The only memory problem I encountered was when I was playing with Haralick features in ``mahotas``, where the [docs](https://mahotas.readthedocs.io/en/latest/faq.html#i-ran-out-of-memory-computing-haralick-features-on-16-bit-images-is-it-not-supported) state that memory issues can be a problem with >8 bit images

I'd rather preserve as much of the original data as possible and only scale down the images where we absolutely have to.

File: tests/test_pre.py (hunk 9 above)
Comment (starcalibre) [inline tests/test_pre.py]: Manually as in.. loaded ``i``, ``j`` and ``k`` into ipython myself and done the averaging, padding, filtering, etc. myself.. yes, that's how I got these values.

File: microscopium/preprocess.py (hunk 2 above)
Comment (starcalibre) [inline microscopium/preprocess.py]: And how about a separate PR to rename ``tlz`` to ``tz`` after this is merged?

Comment (jni) [inline microscopium/preprocess.py]: separate PR = good idea. =)

File: microscopium/preprocess.py (hunk 3 above)
Comment (jni) [inline microscopium/preprocess.py]: you gonna update this? =D

File: microscopium/preprocess.py (hunk 4 above)
Comment (jni) [inline microscopium/preprocess.py]: yes, it would indeed be easier. =D

File: microscopium/preprocess.py (hunk 5 above)
Comment (jni) [inline microscopium/preprocess.py]: The key here is that we shouldn't be sensitive to fluff. =P I think we're not, thanks to the large-radius median filter?

Comment (jni) [inline microscopium/preprocess.py]: ie, in order for a fluff particle to affect the outcome, it would need to take up more than half of the filter disk. Which could potentially happen I think... Hmmmm. Well, the proof is in the pudding, right? You've tried it without stretchlim currently?

File: tests/test_pre.py (hunk 9 above)
Comment (jni) [inline tests/test_pre.py]: ok good. =) For at least one internal value and one border value, though, you should check the result by hand, ie, actually doing the calculation top to bottom.
